common/quotas/global/algorithm/scenarios_test.go

```diff
@@ -0,0 +1,499 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package algorithm
+
+import (
+	"math"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestWeightAlgorithms(t *testing.T) {
+	/*
+		No algorithm can perfectly predict the future, and some issues are discovered only after trying them out on widely-varying production usage.
+
+		So here are some samples of things we've thought of, how they behave, and maybe some comments on how we feel about them.
+		They are re-implemented here in a simplified form, because it's much easier and clearer (and more long-term implement-able) to do so.
+
+		All are bound to a single key across 3 hosts with a simple 15rps target (5 per host when balanced) (3 hosts chosen because there are many easy mistakes with 2).
+	*/
+	const targetRPS = 15
+	max := func(a, b float64) float64 {
+		if a > b {
+			return a
+		}
+		return b
+	}
+	min := func(a, b float64) float64 {
+		if a < b {
+			return a
+		}
+		return b
+	}
+	_, _ = max, min
+
+	roundSchedule := func(low, high int) [][][3]int {
+		rounds := [][3]int{
+			{high, high, high},
+			{low, low, high},
+			{low, high, high},
+			{low, low, high},
+		}
+		// and repeat that cycle like 16x to let it stabilize
+		all := [][][3]int{rounds} // 1
+		all = append(all, rounds) // 2
+		all = append(all, all...) // 4
+		all = append(all, all...) // 8
+		all = append(all, all...) // 16
+		return all
+	}
+
+	// helper to track accept and reject rates.
+	// requests must be an int-float64 (ensured above, by the round schedule)
+	track := func(accepts, rejects *float64, requests int, rps float64) {
+		accepted := min(float64(requests), math.Floor(rps))
+		*accepts += accepted
+		*rejects += float64(requests) - accepted
+	}
+
+	type data struct {
+		// target for the immediate next period of time, as a sample of how target RPSes settle over time
+		targetRPS [3]float64 // rps that should be allowed by the limiter hosts at the end of the last round
```

Comment:
Could you explain to me more about [3] floats—what each one means and how to read it?

Reply:
I suggest using a struct with named fields to make it easier to understand scenarios.